feat: trie cache configuration

jakmeier opened the pull request:
Make the shard cache max total bytes configurable.
Also add separate  configuration for view caches.

This deprecates the old format for configuring cache capacity.
The old format will still work for now but values in the new format
will overwrite any values set in the old format.

Example of the new format, that sets all normal caches to 50MB, aurora's shard to 100MB, shard 3 to 3GB, and view caches to 30MB:

```json
{
  "trie_cache": {
    "default_max_bytes": 50000000,
    "per_shard_max_bytes": {
      "shard1.v1": 10000000,
      "shard3.v1": 3000000000
    }
  },
  "view_trie_cache": {
    "default_max_bytes": 30000000
  }
}
```

resolves #7564

matklad commented on core/store/src/config.rs:
```diff
@@ -176,3 +190,74 @@ impl Default for MigrationSnapshot {
+#[derive(Clone, Copy, Debug, PartialEq, Eq, Hash)]
+pub struct ShardUIdShortSerialization(ShardVersion, ShardId);
+
+impl From<ShardUIdShortSerialization> for ShardUId {
```
I would prefer us to define a reasonable Display & FromStr for ShardUid (and improve Debug along the way).

jakmeier replied:
Hm, then what would be a reasonable general-purpose string format of ShardUId? One that also works as JSON key for the config?
"shardXvY" doesn't feel right anywhere other than when used as a JSON key.

matklad replied:
I would say `"shard1v0"` would be better than our current debug of `"ShardUId { version: 0, shard_id: 1}"`. Though, I'd probably go for something more compact like `s2:1`. 

The rule-of-thumb here is that it's useful to have canonical string representations for vocabulary types, as this simplifies debugging quite a bit. 

Don't have a super strong opinion here!

jakmeier replied:
That makes sense, I'm on board with using FromStr and Display here.
But I was hoping that I can at least avoid a custom serde serialization / deserialization implementation that way. But [DisplayFromStr](https://docs.rs/serde_with/latestserde_with/struct.DisplayFromStr.html) only works for fields. I think I would need it on the struct to make it work in a map.

Am I missing something, or would your suggested solution end up with custom implementation of `Debub`, `Display`, `FromStr`, `serde::Deserialize`, and `serde::Serialize` for ShardUId?

matklad replied:
Yeah, we'd need all those custom impls. But we perhaps can come up with some generic way to impl serde in terms of display/from string (I think we already hard-code it a couple of times somewhere). 

Last time I look at it, I didn't figure a super-nice way to do that, so we might go for a less-nice way (https://docs.rs/serde_with/latest/serde_with/, or a macro_rules).

matklad replied:
>But [DisplayFromStr](https://docs.rs/serde_with/latestserde_with/struct.DisplayFromStr.html) only works for fields.

Ah, I think you are lookin for 

* https://docs.rs/serde_with/latest/serde_with/derive.DeserializeFromStr.html
* https://docs.rs/serde_with/latest/serde_with/derive.SerializeDisplay.html

jakmeier replied:
Ah yes, that's what I was looking for thanks! But looking at it again, it seems that a custom implementation is simpler than pulling in another dependency. (I didn't realize it would be an extra dependency at first.)

I could find at least one more place that could also use `DeserializeFromStr` and probably there are many playes. If we want to use `serde_with` we should probably do it in a separate PR and use it in as many places as possible. But I'm not even sure it would be an improvement.

jakmeier commented on core/primitives/src/shard_layout.rs:
```diff
@@ -355,6 +356,106 @@ pub fn get_block_shard_uid_rev(
+        let (shard_str, version_str) = s
+            .split_once(".")
+            .ok_or_else(|| format!("shard version and number must be separated by \".\""))?;
+        if !version_str.starts_with("v") {
```
using  `strip_prefix` now but not sure if there is a more elegant way...

jakmeier replied:
I guess `map_err` makes it quite a bit cleaner

matklad commented on core/store/src/config.rs:
```diff
@@ -37,12 +38,14 @@ pub struct StoreConfig {
-    /// Default value: ShardUId {version: 1, shard_id: 3} -> 45_000_000
-    /// We're still experimenting with this parameter and it seems decreasing its value can improve
-    /// the performance of the storage
+    /// DEPRECATED: use `trie_cache` instead.
```
When adding a derecation comments, it's a good idea to mention the future release when the stuff should be removed. That way, someone from the future can easily see that the thing can *already* be removed.

matklad commented on core/store/src/config.rs:
```diff
@@ -176,3 +189,11 @@ impl Default for MigrationSnapshot {
+    /// Limit the number trie nodes that fit in the cache.
+    pub max_entries: Option<u64>,
+    /// Limit the sum of all value sizes that fit in the cache.
+    pub max_bytes: Option<u64>,
```
TBH, after the investigation into trie sieze which showed that we have about 100 bytes-per-entry overhead, I am leaning stronger that we should configure *just* the size in bytes, and limit the number of entries by saying that each entry is 100 bytes + val.len() in size. 

This also makes sense from the user's pov: it's clear what 2gig of cache means. IN contrast, 2mil entries is opaque.

jakmeier replied:
@Longarithm What do you think? Would it make sense to only have a configuration for total bytes and drop the number of entries config?

matklad commented on core/primitives/src/shard_layout.rs:
```diff
@@ -355,6 +356,106 @@ pub fn get_block_shard_uid_rev(
+            .split_once(".")
+            .ok_or_else(|| format!("shard version and number must be separated by \".\""))?;
+
+        let version_str = version_str
+            .strip_prefix("v")
+            .ok_or_else(|| format!("shard version must start with \"v\""))?;
+        let version = version_str
+            .parse::<ShardVersion>()
+            .map_err(|e| format!("shard version after \"v\" must be a number, {e}"))?;
```
```suggestion
        let version = version_str
            .strip_prefix("v")
            .ok_or_else(|| format!("shard version must start with \"v\""))?;
            .parse::<ShardVersion>()
            .map_err(|e| format!("shard version after \"v\" must be a number, {e}"))?;
```

?

matklad reviewed (approved):
LGTM, but consider if it is worth to switch to size-only bound (which admitedly inflates the scope of work here quite a bit, but also user-visible configs are annoying to change later, so it makes sense to upfront-design this a bit)?

Longarithm commented:
If we measure the size of each entry as `100 bytes + val.len()` or so, then number of entries can be safely deprecated.

jakmeier commented:
Ok now that #7749 changed the internal behaviour to always limit by memory, as suggested by @matklad, this has become a lot simpler.

Roughly my changes are:

- Removed all fields tracking the number of entries.
- Logic around "num entries" -> "memory limit" calculation is now only done once upfront.
- Removed the extra struct `ShardCacheConfig` and just use `TrieCacheConfig` directly.
- Rearranged the format in `config.json` to make it a bit more user friendly

@matklad I think your earlier approval needs a refresh given the substantial changes

@Longarithm probably makes sense for you to look over this, too, to check I didn't mess up the cache limit calculations and to keep you in the loop of how we configure things now

matklad reviewed (approved):
Yup, new format looks reasonable.

Longarithm commented on core/store/src/trie/trie_storage.rs:
```diff
@@ -233,7 +233,12 @@ pub struct TrieCache(pub(crate) Arc<Mutex<TrieCacheInner>>);
     pub fn new(config: &TrieConfig, shard_uid: ShardUId, is_view: bool) -> Self {
-        let total_size_limit = config.shard_cache_total_size_limit(shard_uid, is_view);
+        let total_size_limit = config
+            .shard_cache_config
```
We should take `view_shard_cache_config` here and below if `is_view` is true, and possibly have a test checking that we take correct values for view and regular caches.

jakmeier replied:
Good catch! Fixed and added a test.
